src/MongoDB.Driver.Core/Core/Authentication/AwsSignatureVersion4.cs

```diff
@@ -0,0 +1,169 @@
+        /// <param name="timestamp">The timestamp.</param>
+        public static void CreateAuthorizationRequest(
+            DateTime dateTime,
+            string accessKeyId,
+            SecureString secretAccessKey,
+            string sessionToken,
+            byte[] salt,
+            string host,
+            out string authorizationHeader,
+            out string timestamp)
+        {
+            var body = "Action=GetCallerIdentity&Version=2011-06-15";
+            var region = GetRegion(host);
+            var service = "sts";
+
+            timestamp = dateTime.ToString("yyyyMMddTHHmmssZ");
+
+            var datestamp = dateTime.ToString("yyyyMMdd");
+
+            var requestHeaders = GetRequestHeaders(
+                body: body,
+                contentType: "application/x-www-form-urlencoded",
+                host: host,
+                timestamp: timestamp,
+                sessionToken: sessionToken,
+                nonce: salt);
+
+            var canonicalHeaders = GetCanonicalHeaders(requestHeaders);
+            var signedHeaders = GetSignedHeaders(requestHeaders);
+
+            var canonicalRequest = string.Join("\n", "POST", "/", "", canonicalHeaders, "", signedHeaders, Hash(body));
+            var algorithm = "AWS4-HMAC-SHA256";
+            var credentialScope = $"{datestamp}/{region}/{service}/aws4_request";
+            var stringToSign = string.Join("\n", algorithm, timestamp, credentialScope, Hash(canonicalRequest));
+            var signature = GetSignature(stringToSign, secretAccessKey, datestamp, region, service);
+
+            authorizationHeader = $"{algorithm} Credential={accessKeyId}/{credentialScope}, SignedHeaders={signedHeaders}, Signature={signature}";
+        }
+
+        //private static methods
+        private static string GetCanonicalHeaders(SortedDictionary<string, string> requestHeaders)
+        {
+            return string.Join("\n", requestHeaders.Select(x => $"{x.Key.ToLowerInvariant()}:{x.Value}"));
+        }
+
+        private static string GetRegion(string host)
+        {
+            if (host == "sts.amazonaws.com")
+            {
+                return "us-east-1";
+            }
+
+            var split = host.Split('.');
+            if (split.Count() > 1)
+            {
+                return split[1];
+            }
+
+            return "us-east-1";
+        }
+
+        private static SortedDictionary<string, string> GetRequestHeaders(
+            string body,
+            string contentType,
+            string host,
+            string timestamp,
+            string sessionToken,
+            byte[] nonce)
+        {
+            var requestHeaders = new SortedDictionary<string, string>
+            {
+                ["Content-Type"] = contentType,
+                ["Content-Length"] = body.Length.ToString(),
+                ["Host"] = host,
+                ["X-Amz-Date"] = timestamp,
+                ["X-MongoDB-GS2-CB-Flag"] = "n",
```

Comment:
Still not sure `X-MongoDB-Optional-Data*` whether this value should be precessed somehow

Reply:
This is optional and other drivers do not include it, so we should be fine.